File: Project1/Python/task3.py
import json
import sys
from operator import add
import hashlib
import binascii
import time
from pyspark import SparkContext,SparkConf

start_time = time.time()
print(start_time)

def count_in_a_partition(iterator):
    yield sum(1 for _ in iterator)

def review_partitioner(iterator):
    #return int(hashlib.md5(iterator.encode('utf-8')).hexdigest(),16)
    return binascii.crc32(iterator[:1].encode('utf-8'))

if __name__ == '__main__':
    reviewFile = sys.argv[1]
    outputFile = sys.argv[2]
    partitionType = sys.argv[3]
    partitionNumber = sys.argv[4]
    n = sys.argv[5]

    results = {}

    conf = SparkConf().setAppName("Task3").setMaster("local[*]")
    sc = SparkContext(conf = conf)

    reviewData = sc.textFile(reviewFile).map(lambda row: json.loads(row))
    #Emit key value pairs or the form: (business_id,1) for each business_id
    reviewRDD = reviewData.map(lambda x: (x['business_id'], 1))

    if partitionType != "default":
        reviewRDD = reviewRDD.partitionBy(int(partitionNumber), review_partitioner)

    results['n_partitions'] = reviewRDD.getNumPartitions()
    results['n_items'] = reviewRDD.mapPartitions(count_in_a_partition).collect()
    print( results['n_items'])


    resultsRDD = reviewRDD.reduceByKey(add)
    results['result'] = resultsRDD.filter(lambda x: x[1] > int(n)).collect()

    with open(outputFile, 'w') as output_file:
        json.dump(results, output_file)
    output_file.close()

File: Project1/Scala/task3.scala
import java.io.{BufferedWriter, File, FileWriter}

import com.fasterxml.jackson.databind.ObjectMapper
import com.fasterxml.jackson.module.scala.DefaultScalaModule
import org.apache.log4j.{Level, Logger}
import org.apache.spark.Partitioner
import org.apache.spark.sql.SparkSession
import org.apache.spark.sql.execution.RDDConversions._
import org.apache.spark.rdd.PairRDDFunctions
import org.json4s.jackson.JsonMethods.{compact, parse}

import scala.collection.mutable

    object task3{
      class custom[V](numParts: Int, elements: Long) extends Partitioner {
        override def numPartitions: Int = numParts

        override def getPartition(key: Any): Int = {
          var k = key.toString.charAt(1)
          return (k.hashCode()%numParts)
        }
      }

    def main(args: Array[String]): Unit = {

      Logger.getLogger("org").setLevel(Level.ERROR)

      val ss = SparkSession.builder().appName("task1").config("spark.master", "local[*]").getOrCreate()
      val sc = ss.sparkContext
      var spark = "cus"
      var n = 20

      var jsonRDD = sc.textFile("/Users/akshatasheth/Desktop/Data Mining/Assignment 1/review.json")
      var reviewRDD = jsonRDD.map { row => (compact(parse(row) \ "business_id"), 1) }
      var finalRDD = reviewRDD

      if (spark != "Default") {
        var finalRDD = reviewRDD.partitionBy(new custom(20, reviewRDD.count()))

      }

      var num_items = finalRDD.mapPartitions(iter => Array(iter.size).iterator, true).collect()
      var results = finalRDD.reduceByKey(_+_).filter( x => x._2 > n).collect()

      val task2 = mutable.LinkedHashMap(
        "n_partition" -> num_items.length,
        "n_items" -> num_items,
      "result" -> results)

      val mapper = new ObjectMapper()
      mapper.registerModule(DefaultScalaModule)

      val file = new File("output2.json")
      val bw = new BufferedWriter(new FileWriter(file))
      bw.write(mapper.writeValueAsString(task2).replaceAll("\\\\\"", ""))
      bw.close()
      //println(m1.toList)



    }
  }

